docs/introduction/framework-extensibility.md

```diff
@@ -78,3 +78,131 @@ as well as a list of customizations that are not (and will not be) possible at a
     * separate users login credentials
     * share company attributes
     * change association from 1:1 to 1:N
+
+## Making the static analysis understand the extended code
+### Problem description
+When extending framework classes, it may happen that tools for static analysis (e.g. PHPStan, PHPStorm) will not understand your code properly.
+Imagine this situation:
+- You have a controller that is dependent on a framework service:
+```php
+namespace Shopsys\ShopBundle\Controller\Front;
+
+use Shopsys\FrameworkBundle\Model\Product\ProductFacade;
+
+class ProductController
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Product\ProductFacade
+     */
+    protected $productFacade;
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\ProductFacade $productFacade
+     */
+    public function __construct(ProductFacade $productFacade)
+    {
+        $this->productFacade = $productFacade;
+    }
+}
+```
+- In your project, you extend the framework service:
+```php
+namespace Shopsys\ShopBundle\Model\Product;
+
+use Shopsys\FrameworkBundle\Model\Product\ProductFacade as BaseProductFacade;
+
+class ProductFacade extends BaseProductFacade
+{
+    public function myCustomAwesomeFunction()
+    {
+        return 42;
+    }
+}
+```
+- You register your extension in DI services configuration and thanks to that, your class is used in `ProductController` instead of the one from `FrameworkBundle`, so far so good:
+```yaml
+Shopsys\FrameworkBundle\Model\Product\ProductFacade: '@Shopsys\ShopBundle\Model\Product\ProductFacade'
+```
+**However, when you want to use your `myCustomAwesomeFunction()` in `ProductController`, the static analysis is not aware of that function.**
+#### Solution
+To fix this, you need to change the annotations properly:
+```diff
+class ProductController
+{
+      /**
+-      * @var \Shopsys\FrameworkBundle\Model\Product\ProductFacade
++      * @var \Shopsys\ShopBundle\Model\Product\ProductFacade
+       */
+      protected $productFacade;
+
+      /**
+-      * @param \Shopsys\FrameworkBundle\Model\Product\ProductFacade $productFacade
++      * @param \Shopsys\ShopBundle\Model\Product\ProductFacade $productFacade
+       */
+      public function __construct(ProductFacade $productFacade)
+      {
+          $this->productFacade = $productFacade;
+      }
+}
+```
+**Luckily, you do need to fix the annotations manually, there is the [Phing target `fix-annotations-in-extended-classes`](./console-commands-for-application-management-phing-targets.md#fix-annotations-in-extended-classes), that handles everything for you.**
```

Comment:
Could this name be longer? :D ... It really can't be called just `fix-annotations`? :)

Reply:
thanks for your input, changed :wink: